File: async_checker/optimized/async_checker.nim
type
  TaskResult = object
    latency: int
    status: int

proc runTask(id: int): TaskResult {.inline.} =
  let seed = ((uint64(id) * 1664525'u64 + 1013904223'u64) and 0xFFFFFFFF'u64).uint32
  let latency = int(10 + (seed mod 990))
  var status = 200
  if (seed mod 7) == 0:
    status = if (seed mod 3) == 0: 429 else: 500
  result = TaskResult(latency: latency, status: status)

proc main() =
  let tasks = 10000
  var okCount = 0
  var rlCount = 0
  var errCount = 0
  var totalLat = 0'i64
  var histogram: array[1000, int]

  for i in 0 ..< tasks:
    let res = runTask(i)
    inc(histogram[res.latency])
    totalLat += res.latency
    case res.status
    of 200: inc(okCount)
    of 429: inc(rlCount)
    of 500: inc(errCount)
    else: discard

  var count = 0
  var p50, p95, p99 = 0
  let targetP50 = int(float(tasks) * 0.50)
  let targetP95 = int(float(tasks) * 0.95)
  let targetP99 = int(float(tasks) * 0.99)

  for lat in 0 ..< 1000:
    let c = histogram[lat]
    if c > 0:
      if count < targetP50 and count + c >= targetP50: p50 = lat
      if count < targetP95 and count + c >= targetP95: p95 = lat
      if count < targetP99 and count + c >= targetP99: p99 = lat
      count += c

  echo "Async complete: tasks=", tasks, ", ok=", okCount, ", rate_limited=", rlCount, ", errors=", errCount, ", latency_sum=", totalLat, ", p50=", p50, ", p95=", p95, ", p99=", p99

main()

File: async_checker/optimized/async_checker.rs
use std::sync::mpsc;
use std::thread;

struct TaskResult {
    latency: usize,
    status: usize,
}

#[inline(always)]
fn run_task(id: usize) -> TaskResult {
    let seed = ((id as u64).wrapping_mul(1664525) + 1013904223) & 0xFFFFFFFF;
    let latency = (10 + (seed % 990)) as usize;
    let status = if (seed % 7) != 0 {
        200
    } else if (seed % 3) == 0 {
        429
    } else {
        500
    };
    TaskResult { latency, status }
}

fn main() {
    let tasks = 10000;
    let workers = 16;
    let (tx, rx) = mpsc::sync_channel::<TaskResult>(1000);

    let chunk_size = tasks / workers;
    let mut handles = Vec::new();

    for w in 0..workers {
        let tx = tx.clone();
        let start = w * chunk_size;
        let end = if w == workers - 1 { tasks } else { start + chunk_size };
        handles.push(thread::spawn(move || {
            for i in start..end {
                let _ = tx.send(run_task(i));
            }
        }));
    }
    drop(tx);

    let mut ok_count = 0;
    let mut rl_count = 0;
    let mut err_count = 0;
    let mut total_lat = 0u64;
    let mut histogram = [0usize; 1000];

    for res in rx {
        histogram[res.latency] += 1;
        total_lat += res.latency as u64;
        match res.status {
            200 => ok_count += 1,
            429 => rl_count += 1,
            500 => err_count += 1,
            _ => {}
        }
    }

    for h in handles {
        let _ = h.join();
    }

    let mut count = 0;
    let mut p50 = 0;
    let mut p95 = 0;
    let mut p99 = 0;
    let target_p50 = (tasks as f64 * 0.50) as usize;
    let target_p95 = (tasks as f64 * 0.95) as usize;
    let target_p99 = (tasks as f64 * 0.99) as usize;

    for (lat, &c) in histogram.iter().enumerate() {
        if c > 0 {
            if count < target_p50 && count + c >= target_p50 {
                p50 = lat;
            }
            if count < target_p95 && count + c >= target_p95 {
                p95 = lat;
            }
            if count < target_p99 && count + c >= target_p99 {
                p99 = lat;
            }
            count += c;
        }
    }

    println!(
        "Async complete: tasks={}, ok={}, rate_limited={}, errors={}, latency_sum={}, p50={}, p95={}, p99={}",
        tasks, ok_count, rl_count, err_count, total_lat, p50, p95, p99
    );
}
